File: app/OnyxPenDemo/src/main/java/com/onyx/android/eink/pen/demo/erase/bean/EraseContext.java
package com.onyx.android.eink.pen.demo.erase.bean;

import android.graphics.RectF;

import com.onyx.android.eink.pen.demo.brush.shape.Shape;
import com.onyx.android.sdk.data.note.TouchPoint;
import com.onyx.android.sdk.pen.data.TouchPointList;

import java.util.ArrayList;
import java.util.List;

public class EraseContext {
    private final List<Shape> splitShapes = new ArrayList<>();
    private final TouchPointList wholeEraseTrackPoints = new TouchPointList();
    private RectF eraseRect;
    private volatile boolean finishing;

    public boolean isFinishing() {
        return finishing;
    }

    public void setFinishing(boolean finishing) {
        this.finishing = finishing;
    }

    /** Original shapes removed by split (used for pen-up dirty rect union). */
    public List<Shape> getSplitShapes() {
        return splitShapes;
    }

    public TouchPointList getWholeEraseTrackPoints() {
        return wholeEraseTrackPoints;
    }

    public void addErasePoint(TouchPoint point) {
        wholeEraseTrackPoints.add(point);
    }

    public void addErasePoints(TouchPointList pointList) {
        if (pointList == null) {
            return;
        }
        wholeEraseTrackPoints.addAll(pointList);
    }

    public void addSplitShapes(List<Shape> shapes) {
        splitShapes.addAll(shapes);
    }

    public void unionEraseRect(RectF rect) {
        if (rect == null) {
            return;
        }
        if (eraseRect == null) {
            eraseRect = new RectF(rect);
        } else {
            eraseRect.union(rect);
        }
    }

    public RectF getEraseRect() {
        return eraseRect;
    }
}

File: app/OnyxPenDemo/src/main/java/com/onyx/android/eink/pen/demo/erase/bean/EraseContext.kt
package com.onyx.android.eink.pen.demo.erase.bean

import android.graphics.RectF
import com.onyx.android.eink.pen.demo.shape.Shape
import com.onyx.android.sdk.data.note.TouchPoint
import com.onyx.android.sdk.pen.data.TouchPointList
import kotlin.concurrent.Volatile

class EraseContext {
    private val splitShapes: MutableList<Shape?> = ArrayList()
    private val wholeEraseTrackPoints: TouchPointList = TouchPointList()
    private var eraseRect: RectF? = null

    @Volatile
    private var finishing: Boolean = false

    fun isFinishing(): Boolean = finishing

    fun setFinishing(finishing: Boolean) {
        this.finishing = finishing
    }

    fun getSplitShapes(): MutableList<Shape?> = splitShapes

    fun getWholeEraseTrackPoints(): TouchPointList = wholeEraseTrackPoints

    fun addErasePoint(point: TouchPoint?) {
        wholeEraseTrackPoints.add(point)
    }

    fun addErasePoints(pointList: TouchPointList?) {
        if (pointList == null) {
            return
        }
        wholeEraseTrackPoints.addAll(pointList)
    }

    fun addSplitShapes(shapes: MutableList<Shape?>) {
        splitShapes.addAll(shapes)
    }

    fun unionEraseRect(rect: RectF?) {
        if (rect == null) {
            return
        }
        val current = eraseRect
        if (current == null) {
            eraseRect = RectF(rect)
        } else {
            current.union(rect)
        }
    }

    fun getEraseRect(): RectF? = eraseRect
}
